text/0126-introduce-xcq.md

```diff
@@ -0,0 +1,384 @@
+- XCQ query arguments: Query arguments like accounts to be queried are also passed together with the query program.
+
+### XCQ Runtime API
+
+The runtime API for off-chain query usage includes two methods:
+
+- `execute_query`: Executes the query and returns the result. It takes the query, input, and weight limit as arguments.
+The query is the query program in PolkaVM program binary format.
+The input is the query arguments that is SCALE-encoded.
+The weight limit is the maximum weight allowed for the query execution.
+- `metadata`: Return metadata of supported extensions (introduced in later section) and methods, serving as a feature discovery functionality
+
+**Example XCQ Runtime API**:
+```rust
+decl_runtime_apis! {
+    pub trait XcqApi {
+        fn execute_query(query: Vec<u8>, input: Vec<u8>, weight_limit: u64) -> XcqResult;
+        fn metadata() -> Vec<u8>;
+    }
+}
+type XcqResult =  Result<XcqResponse, XcqError>;
+type XcqResponse = Vec<u8>;
+enum XcqError {
+    Custom(String),
+}
+```
+
+**Example metadata**:
+```rust
+Metadata {
+    extensions: vec![
+        ExtensionMetadata {
+            name: "ExtensionCore",
+            methods: vec![MethodMetadata {
+                name: "has_extension",
+                inputs: vec![MethodParamMetadata {
+                    name: "id",
+                    ty: XcqType::Primitive(PrimitiveType::U64)
+                }],
+                output: XcqType::Primitive(PrimitiveType::Bool)
+            }]
+        },
+        ExtensionMetadata {
+            name: "ExtensionFungibles",
+            methods: vec![
+                MethodMetadata {
+                    name: "total_supply",
+                    inputs: vec![MethodParamMetadata {
+                        name: "asset",
+                        ty: XcqType::Primitive(PrimitiveType::U32)
+                    }],
+                    output: XcqType::Primitive(PrimitiveType::U64)
+                },
+                MethodMetadata {
+                    name: "balance",
+                    inputs: vec![
+                        MethodParamMetadata {
+                            name: "asset",
+                            ty: XcqType::Primitive(PrimitiveType::U32)
+                        },
+                        MethodParamMetadata {
+                            name: "who",
+                            ty: XcqType::Primitive(PrimitiveType::H256)
+                        }
+                    ],
+                    output: XcqType::Primitive(PrimitiveType::U64)
+                }
+            ]
+        }
+    ]
+}
+```
+
+Note: `ty` is represented by a meta-type system called `xcq-types`
+
+#### xcq-types
+
+`xcq-types` is a meta-type system similar to `scale-info` but simpler.
+It enables different chains with different type definitions to work via a common operation.
+Front-end codes constructs call data to XCQ programs according to metadata provided by different chains.
```

Comment:
I'd like to know more about `xcq-types`, ie what are the existing issues with `scale-info` here? Since it just describes the shapes of types, one can compare two sets of `scale-info` types across chains already, or is there more to it?

Reply:
Another point here: I'd assume that we would want XCQ metadata to be a part of the frame-metadata that we hand back when users call the `Metadata_metadata_at_version` runtime API? In that case, it feels beneficial to use the existing `scale_info` based type information used by the rest of the metadata over having a separate concept to understand and integrate there.

Reply:
Yes, makes sense. we are experimenting with switching back to `scale-info`.